### PR TITLE
WIP: low-level code for HW wallet PSBT integration

### PR DESCRIPTION
**Pull request by Overtorment**
wip. do not merge

### CODE REVIEW

File: class/hd-segwit-bech32-wallet.js [hunk 1]
```diff
@@ -650,6 +675,29 @@ export class HDSegwitBech32Wallet extends AbstractHDWallet {
     return this._utxo;
   }
 
+  _getDerivationPathByAddress(address) {
+    const path = "m/84'/0'/0'";
```
Comment (stepansnigirev) [inline class/hd-segwit-bech32-wallet.js]: I suggest to treat the parent of zpub as a master key, and use derivation path of `m/zpub.child_index`
If in zpub's parent fingerprint is 00000000 this means that zpub is root - in this case we should use derivation path of just `m/`
(i.e. Bitcoin Core uses this type of derivation, as far as I remember they don't use accounts and bip84)

File: class/hd-segwit-bech32-wallet.js [hunk 2]
```diff
@@ -737,6 +785,94 @@ export class HDSegwitBech32Wallet extends AbstractHDWallet {
...
+    }
+
+    let psbt = new bitcoin.Psbt();
+
+    let c = 0;
+    let values = {};
+
+    inputs.forEach(input => {
+      let pubkey = this._getPubkeyByAddress(input.address);
+      let masterFingerprint = Buffer.from([0x00, 0x00, 0x00, 0x00]);
```
Comment (stepansnigirev) [inline class/hd-segwit-bech32-wallet.js]: here we can use parent fingerprint from zpub

File: class/hd-segwit-bech32-wallet.js [hunk 3]
```diff
@@ -737,6 +785,94 @@ export class HDSegwitBech32Wallet extends AbstractHDWallet {
...
+      // if output has no address - this is change output
+      let change = false;
+      if (!output.address) {
+        output.address = changeAddress;
+        change = true;
+      }
+
+      let path = this._getDerivationPathByAddress(output.address);
+      let pubkey = this._getPubkeyByAddress(output.address);
+      let masterFingerprint = Buffer.from([0x00, 0x00, 0x00, 0x00]);
```
Comment (stepansnigirev) [inline class/hd-segwit-bech32-wallet.js]: here we can use parent fingerprint from zpub

File: tests/integration/hd-segwit-bech32-wallet.test.js [hunk 4]
```diff
@@ -40,6 +40,11 @@ describe('Bech32 Segwit HD (BIP84)', () => {
     assert.strictEqual(hd._getInternalAddressByIndex(0), 'bc1q8c6fshw2dlwun7ekn9qwf37cu2rn755upcp6el');
     assert.ok(hd._getInternalAddressByIndex(0) !== hd._getInternalAddressByIndex(1));
 
+    assert.strictEqual(hd._getDerivationPathByAddress(hd._getExternalAddressByIndex(0)), "m/84'/0'/0'/0/0");
```
Comment (stepansnigirev) [inline tests/integration/hd-segwit-bech32-wallet.test.js]: with comments above this should be `m/0'/0/0`
everything below as well

**Comment by Overtorment** [comment]
Created via API

**Comment by Overtorment** [comment]
@dr3am3r can you pls reviev?

**Comment by Overtorment** [comment]
Test this build in browser:

(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧ https://appetize.io/app/ceytb68f4r1ezy23e7eb5ww5v8

(posted automatically)

**Comment by Overtorment** [comment]
Test this build in browser:

(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧ https://appetize.io/app/ex63bz0xk9x1pf2dfjumntb7bm

(posted automatically)

**Comment by Overtorment** [comment]
Test this build in browser:

(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧ https://appetize.io/app/t763wvv049enf9kcv50dt3hj14

(posted automatically)

File: screen/wallets/details.js [hunk 5]
```diff
@@ -92,6 +102,12 @@ export default class WalletDetails extends Component {
     }
   }
 
+  async onUseWithHardwareWalletSwitch(value) {
+    let wallet = this.state.wallet;
+    wallet.use_with_hardware_wallet = !!value;
+    this.setState({ useWithHardwareWallet: !!value, wallet });
```
Comment (dr3am3r) [inline screen/wallets/details.js]: when relying on previous state's value should use updater function instead, see 
https://en.reactjs.org/docs/react-component.html#setstate

File: screen/send/psbtWithHardwareWallet.js [hunk 6]
```diff
@@ -0,0 +1,250 @@
...
+        <View style={{ flex: 1, paddingTop: 20 }}>
+          <ActivityIndicator />
+        </View>
+      );
+    }
+
+    if (this.state.success) return this._renderSuccess();
+    if (this.state.renderScanner) return this._renderScanner();
+    if (this.state.txhex) return this._renderBroadcastHex();
+
```
Comment (dr3am3r) [inline screen/send/psbtWithHardwareWallet.js]: `switch .... case... ` would look much cleaner

File: screen/send/psbtWithHardwareWallet.js [hunk 7]
```diff
@@ -0,0 +1,250 @@
...
+          />
+
+          <TouchableOpacity style={{ marginVertical: 24 }} onPress={() => Clipboard.setString(this.state.txhex)}>
+            <Text style={{ color: '#9aa0aa', fontSize: 15, fontWeight: '500', alignSelf: 'center' }}>Copy and broadcast later</Text>
+          </TouchableOpacity>
+          <TouchableOpacity style={{ marginVertical: 24 }} onPress={() => Linking.openURL('https://coinb.in/?verify=' + this.state.txhex)}>
+            <Text style={{ color: '#9aa0aa', fontSize: 15, fontWeight: '500', alignSelf: 'center' }}>Verify on coinb.in</Text>
+          </TouchableOpacity>
+          <BlueSpacing20 />
+          <BlueButton onPress={() => this.broadcast()} title={loc.send.confirm.sendNow} />
```
Comment (dr3am3r) [inline screen/send/psbtWithHardwareWallet.js]: no need for an extra closure, simply
`onPress={this.broadcast}`

File: screen/send/psbtWithHardwareWallet.js [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+            marginBottom: 53,
+          }}
+        >
+          <Icon name="check" size={50} type="font-awesome" color="#0f5cc0" />
+        </View>
+        <BlueCard>
+          <BlueButton
+            onPress={() => {
+              this.props.navigation.dismiss();
+            }}
```
Comment (dr3am3r) [inline screen/send/psbtWithHardwareWallet.js]: no need for an extra closure, simply
`onPress={this.props.navigation.dismiss}`

**Comment by Overtorment** [comment]
Test this build in browser:

(ﾉ◕ヮ◕)ﾉ*:・ﾟ✧ https://appetize.io/app/6w0m9kf3jj7zh859gxzc0erdw0

(posted automatically)

**Pull request merged**